File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans;
        vector<int> pref(n , 1) , suff(n , 1);
        int prod = nums[0];
        for(int i = 1 ; i < n ; i++){
            pref[i] = prod;
            prod *= nums[i];
        }
        prod = nums[n - 1];
        for(int i = n - 2 ; i >= 0 ; i--){
            suff[i] = prod;
            prod *= nums[i];
        }

        for(int i = 0 ; i< n ; i++){
            ans.push_back(pref[i] * suff[i]);
        }
        return ans;
    } 
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        n = len(nums)
        ans = [1] * n
        prod = nums[0]
        for i in range(1 , n):
            ans[i] = prod
            prod *= nums[i]
        
        prod = nums[n - 1]
        for i in range(n - 2 , -1 , -1):
            ans[i] = ans[i] * prod
            prod *= nums[i]
        return ans
        
        